src/read.c

```diff
@@ -4688,37 +4730,23 @@ static avifResult avifReadColorProperties(avifIO * io,
 {
     // Find and adopt all colr boxes "at most one for a given value of colour type" (HEIF 6.5.5.1, from Amendment 3)
     // Accept one of each type, and bail out if more than one of a given type is provided.
+    AVIF_CHECKRES(avifReadColorNclxProperty(properties, colorPrimaries, transferCharacteristics, matrixCoefficients, yuvRange, cicpSet));
     avifBool colrICCSeen = AVIF_FALSE;
-    avifBool colrNCLXSeen = AVIF_FALSE;
     for (uint32_t propertyIndex = 0; propertyIndex < properties->count; ++propertyIndex) {
         avifProperty * prop = &properties->prop[propertyIndex];
-
-        if (!memcmp(prop->type, "colr", 4)) {
-            if (prop->u.colr.hasICC) {
-                if (colrICCSeen) {
-                    return AVIF_RESULT_BMFF_PARSE_FAILED;
-                }
-                avifROData iccRead;
-                AVIF_CHECKRES(io->read(io, 0, prop->u.colr.iccOffset, prop->u.colr.iccSize, &iccRead));
-                colrICCSeen = AVIF_TRUE;
-                AVIF_CHECKRES(avifRWDataSet(icc, iccRead.data, iccRead.size));
-            }
-            if (prop->u.colr.hasNCLX) {
-                if (colrNCLXSeen) {
-                    return AVIF_RESULT_BMFF_PARSE_FAILED;
+        if (!memcmp(prop->type, "colr", 4) && prop->u.colr.hasICC) {
+            if (colrICCSeen) {
+                if (icc->size > 0) {
+                    avifRWDataFree(icc);
```

Comment:
It is not enogh to free `icc` here, because this function may also return an error from line 4745 or line 4747 (via the `AVIF_CHECKRES` macro).

It is better to just document that the caller needs to free `icc` whether this functions succeeds or fails at line 4721, and then verify that the callers do that. I verified that is the case because `icc` will eventually be freed when we destroy the decoder.

Reply:
Done, i have updated the comment and reverted the latest change.